src/main/rx/ghst.c

```diff
@@ -0,0 +1,259 @@
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "platform.h"
+
+#ifdef USE_SERIALRX_GHST
+
+#include "build/build_config.h"
+#include "build/debug.h"
+
+#include "common/crc.h"
+#include "common/maths.h"
+#include "common/utils.h"
+
+#include "pg/rx.h"
+
+#include "drivers/serial.h"
+#include "drivers/serial_uart.h"
+#include "drivers/system.h"
+#include "drivers/time.h"
+
+#include "io/serial.h"
+
+#include "rx/rx.h"
+#include "rx/ghst.h"
+
+#include "telemetry/ghst.h"
+
+// NOTE: SERIAL_BIDIR appears to use OD drive on telemetry data from FC to Rx (slow rise times)
+// SERIAL_BIDIR_PP would appear to be the correct way to do this, but this mode kills incoming control data
+#define GHST_PORT_OPTIONS               (SERIAL_STOPBITS_1 | SERIAL_PARITY_NO | SERIAL_BIDIR)		
+#define GHST_PORT_MODE                  MODE_RXTX   // bidirectional on single pin
+
+#define GHST_MAX_FRAME_TIME_US          500         // 14 bytes @ 420k = ~450us
+#define GHST_TIME_BETWEEN_FRAMES_US     4500        // fastest frame rate = 222.22Hz, or 4500us
+
+#define GHST_PAYLOAD_OFFSET offsetof(ghstFrameDef_t, type)
+
+STATIC_UNIT_TESTED bool ghstFrameDone = false;
+STATIC_UNIT_TESTED ghstFrame_t ghstFrame;
+STATIC_UNIT_TESTED ghstFrame_t ghstChannelDataFrame;
+STATIC_UNIT_TESTED uint32_t ghstChannelData[GHST_MAX_NUM_CHANNELS];
+
+static serialPort_t *serialPort;
+static timeUs_t ghstFrameStartAtUs = 0;
+static uint8_t telemetryBuf[GHST_FRAME_SIZE_MAX];
+static uint8_t telemetryBufLen = 0;
+
+static timeUs_t lastRcFrameTimeUs = 0;
+
+/* GHST Protocol
+ * Ghost uses 420k baud single-wire, half duplex connection, connected to a FC UART 'Tx' pin
+ * Each control packet is interleaved with one or more corresponding downlink packets
+ * 
+ * Uplink packet format (Control packets)
+ * <Addr><Len><Type><Payload><CRC>
+ *
+ * Addr:        u8      Destination address
+ * Len          u8      Length includes the packet ID, but not the CRC
+ * CRC          u8
+ * 
+ * Ghost packets are designed to be as short as possible, for minimum latency.
+ * 
+ * Note that the GHST protocol does not handle, itself, failsafe conditions. Packets are passed from 
+ * the Ghost receiver to Betaflight as and when they arrive. Betaflight itself is responsible for
+ * determining when a failsafe is necessary based on dropped packets. 
+ * 
+  */
+
+#define GHST_FRAME_LENGTH_ADDRESS       1
+#define GHST_FRAME_LENGTH_FRAMELENGTH   1
+#define GHST_FRAME_LENGTH_TYPE_CRC      1
+
+STATIC_UNIT_TESTED uint8_t ghstFrameCRC(void)
+{
+    // CRC includes type and payload
+    uint8_t crc = crc8_dvb_s2(0, ghstFrame.frame.type);
+    for (int i = 0; i < ghstFrame.frame.len - GHST_FRAME_LENGTH_TYPE_CRC - 1; ++i) {
+        crc = crc8_dvb_s2(crc, ghstFrame.frame.payload[i]);
+    }
+    return crc;
+}
+
+// Receive ISR callback, called back from serial port
+STATIC_UNIT_TESTED void ghstDataReceive(uint16_t c, void *data)
```

Comment:
This should be refactored:
- the ISR (which runs outside of any scheduling) should only contain the minimum amount of code to put the incoming data into a buffer;
- `rcFrameStatusFn` (which runs at high frequency outside of the normal task scheduling) should only contain the minimum amount of code to check if a complete frame has been received, and indicate this by return value;
- `rcProcessFrameFn` (which is scheduled as a task every time frame completion has been indicated) should do the decoding of the incoming data and all other required functions.
See `fport.c` for an example.

Reply:
Ok, took a while but I believe the just-submitted change does what you propose.